crates/scope/src/pass.rs

```diff
@@ -122,4 +113,55 @@ impl<'alloc> Pass<'alloc> for ScopePass<'alloc> {
         self.context.before_function_declaration(name, ast);
         self.function_map.insert(ast, ast);
     }
+
+    fn enter_enum_expression_variant_function_expression(&mut self, ast: &'alloc Function<'alloc>) {
+        self.context.before_function_expression(ast);
+    }
+
+    fn leave_enum_expression_variant_function_expression(
+        &mut self,
+        _ast: &'alloc Function<'alloc>,
+    ) {
+        self.context.after_function_expression();
+    }
+
+    fn visit_formal_parameters(&mut self, ast: &'alloc FormalParameters<'alloc>) {
+        self.context.before_function_parameters(ast);
```

Comment:
another possibility is that we do both at the same time, with transaction data.

so, when we hit `(':
  1. create transaction data for:
     * A. scope analysis for expression in parenthesis
     * B. scope analysis for arrow function parameter
  2. while parsing the content of `(...)`:
     * A. note identifiers as use
     * B. note identifiers as parameter
  3. after `)`:
     * when hit `=>`
        * discard A
        * commit B and keep scope analysis for function
     * otherwise
        * discard B
        * commit A to enclosing scope

maybe we could do the same for early errors, with storing pending errors in the transaction.

the problem is that the transaction can be nested, and handling it might be a nightmare :P